lucene/core/src/java/org/apache/lucene/index/SegmentInfo.java

```diff
@@ -111,12 +112,12 @@ public SegmentInfo(Directory dir, Version version, Version minVersion, String na
     this.maxDoc = maxDoc;
     this.isCompoundFile = isCompoundFile;
     this.codec = codec;
-    this.diagnostics = Objects.requireNonNull(diagnostics);
+    this.diagnostics = Collections.unmodifiableMap(new HashMap<>(Objects.requireNonNull(diagnostics)));
     this.id = id;
     if (id.length != StringHelper.ID_LENGTH) {
       throw new IllegalArgumentException("invalid id: " + Arrays.toString(id));
     }
-    this.attributes = Objects.requireNonNull(attributes);
+    this.attributes = Collections.unmodifiableMap(new HashMap<>(Objects.requireNonNull(attributes)));
```

Comment:
Nit: now that we are copying the input attributes, we should change the few call sites that pass a `new HashMap<>()` as attributes to pass a `Collections.emptyMap()` instead?

Reply:
do we have some? I did a quick look and I didn't find any

Reply:
IndexWriter#mergeInit, IndexWriter#addIndexes, IndexSplitter#split seem to do that

Reply:
One thing: once master is on Java 11, this should use: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Map.html#copyOf(java.util.Map)

We may simplify more code doing that as Java 11 has now unmodifiable maps using a static initializer like Guava once had with ImmutableMap. Same for lists and sets.

Reply:
++ I will be able to help here